metrics/chronos_httpserver.nim

```diff
@@ -55,10 +78,367 @@ when defined(metrics):
+
+    let
+      http =
+        block:
+          let socketFlags = {ServerFlags.ReuseAddr}
+          let res = HttpServerRef.new(server.address, cb,
+                                      socketFlags = socketFlags)
+          if res.isErr():
+            lastError = res.error()
+            nil
+          else:
+            res.get()
+      reqTransp =
+        try:
+          fromPipe(server.requestPipe.read)
+        except CatchableError as exc:
+          await http.closeWait()
+          return
+      respTransp =
+        try:
+          fromPipe(server.responsePipe.write)
+        except CatchableError as exc:
+          await http.closeWait()
+          await reqTransp.closeWait()
+          return
+      threadData = MetricsThreadData(
+        reqTransp: reqTransp, respTransp: respTransp, http: http)
+
+    while true:
+      let res = await asyncStep(server, threadData, lastError)
+      if not(res):
+        break
+
+    await allFutures(reqTransp.closeWait(), respTransp.closeWait())
+
+  proc serveMetricsServer(server: MetricsServerData) {.thread.} =
+    ignoreSignalsInThread()
+    let loop {.used.} = getThreadDispatcher()
+    try:
+      waitFor asyncLoop(server)
+    except Exception as exc:
+      printError($exc.msg)
 
 proc startMetricsHttpServer*(address = "127.0.0.1", port = Port(8000)) {.
-     raises: [Exception].} =
+     raises: [Exception],
+     deprecated: "Please use MetricsHttpServerRef API".} =
   when defined(metrics):
     httpServerThread.createThread(serveHttp, initTAddress(address, port))
 
+proc new*(t: typedesc[MetricsHttpServerRef], address: string,
+          port: Port): Result[MetricsHttpServerRef, cstring] {.
+     raises: [Defect].} =
+  ## Initialize new instance of MetricsHttpServerRef.
+  ##
+  ## This involves creation of new thread and new processing loop in the new
+  ## thread.
+  when defined(metrics):
+    template closePipe(b: untyped): untyped =
+      closeHandle(b.read)
+      closeHandle(b.write)
+    let
+      taddress =
+        try:
+          initTAddress(address, port)
+        except TransportAddressError as exc:
+          return err("Invalid server address")
+      request =
+        block:
+          let res = createAsyncPipe()
+          if (res.read == asyncInvalidPipe) or (res.write == asyncInvalidPipe):
+            return err("Unable to create communication request pipe")
+          res
+      response =
+        block:
+          let res = createAsyncPipe()
+          if (res.read == asyncInvalidPipe) or (res.write == asyncInvalidPipe):
+            request.closePipe()
+            return err("Unable to create communication response pipe")
+          res
+      data = MetricsServerData(address: taddress, requestPipe: request,
+                               responsePipe: response)
+    var server = MetricsHttpServerRef(data: data)
+    try:
+      createThread(server.thread, serveMetricsServer, data)
+    except Exception as exc:
+      request.closePipe()
```

Comment:
`defer` or `finally` will be executed on error and on success, but i wish to close pipes only on failure.

Reply:
Yes, but you can easily add an extra `bool` variable that will guard the `close` call in the finally section and you would disarm this bool variable at the very last line of the function (in the success path). This is similar to how in C++, you can allocate memory inside a function, store the allocated buffer in a smart pointer such as `auto_ptr` and then call `return ptr.release()` at the last line of the function (the "disarming" variable is then part of the smart pointer).

Reply:
@zah even in such case usage of `finally` will require to create nested `try` block. also its not so flexible as you, because i can't reuse single `try/finally` for all my error handling.

For example i could use it only for this part of code and i dont think it looks great:

```nim
    var success = false
    try:
      try:
        createThread(server.thread, serveMetricsServer, data)
      except Exception as exc:
        return err("Unexpected error while spawning metrics server's thread")
      except ResourceExhaustedError as exc:
        return err("Unable to spawn metrics server's thread")

      server.reqTransp =
        try:
          fromPipe(request.write)
        except CatchableError as exc:
          return err("Unable to establish communication channel with " &
                     "metrics server thread")
      server.respTransp =
        try:
          fromPipe(response.read)
        except CatchableError as exc:
          return err("Unable to establish communication channel with " &
                     "metrics server thread")
      success = true
      ok(server)
    finally:
      if not(success):
        request.closePipe()
        response.closePipe()
```